code/data_structures/src/tree/binary_tree/avl_tree/avl__tree.py

```diff
@@ -0,0 +1,262 @@
+                    self.node.right.update_balances()
+
+            self.balance = self.node.left.height - self.node.right.height 
+        else: 
+            self.balance = 0 
+
+    def delete(self, key):
+        # debug("Trying to delete at node: " + str(self.node.key))
+        if self.node != None: 
+            if self.node.key == key: 
+                debug("Deleting ... " + str(key))  
+                if self.node.left.node == None and self.node.right.node == None:
+                    self.node = None # leaves can be killed at will 
+                # if only one subtree, take that 
+                elif self.node.left.node == None: 
+                    self.node = self.node.right.node
+                elif self.node.right.node == None: 
+                    self.node = self.node.left.node
+                
+                # worst-case: both children present. Find logical successor
+                else:  
+                    replacement = self.logical_successor(self.node)
+                    if replacement != None: # sanity check 
+                        debug("Found replacement for " + str(key) + " -> " + str(replacement.key))  
+                        self.node.key = replacement.key 
+                        
+                        # replaced. Now delete the key from right child 
+                        self.node.right.delete(replacement.key)
+                    
+                self.rebalance()
+                return  
+            elif key < self.node.key: 
+                self.node.left.delete(key)  
+            elif key > self.node.key: 
+                self.node.right.delete(key)
+                        
+            self.rebalance()
+        else: 
+            return 
+
+    def logical_predecessor(self, node):
+        ''' 
+        Find the biggest valued node in LEFT child
+        ''' 
+        node = node.left.node 
+        if node != None: 
+            while node.right != None:
+                if node.right.node == None: 
+                    return node 
+                else: 
+                    node = node.right.node  
+        return node 
+    
+    def logical_successor(self, node):
+        ''' 
+        Find the smallese valued node in RIGHT child
+        ''' 
+        node = node.right.node  
+        if node != None: # just a sanity check  
+            
+            while node.left != None:
+                debug("LS: traversing: " + str(node.key))
+                if node.left.node == None: 
+                    return node 
+                else: 
+                    node = node.left.node  
+        return node 
+
+    def check_balanced(self):
+        if self == None or self.node == None: 
+            return True
+        
+        # We always need to make sure we are balanced 
+        self.update_heights()
+        self.update_balances()
+        return ((abs(self.balance) < 2) and self.node.left.check_balanced() and self.node.right.check_balanced())  
```

Comment:
Please shorten this line to < 80 chars.

Reply:
Didn't get it.

Reply:
What do you mean?

Reply:
I don't get what to shorten.

Reply:
Try something like this:

```suggestion
        return (
            abs(self.balance) < 2 and 
            self.node.left.check_balanced() and 
            self.node.right.check_balanced())  
```